<small><font color=gray>Notebook author: <a href="https://www.linkedin.com/in/olegmelnikov/" target="_blank">Oleg Melnikov</a> ©2021 onwards</font></small><hr style="margin:0;background-color:silver">

**[<font size=6>🚗Auto</font>](https://www.kaggle.com/competitions/13nov23hse-auto/rules)**. [**Instructions**](https://colab.research.google.com/drive/1riOGrE_Fv-yfIbM5V4pgJx4DWcd92cZr#scrollTo=ITaPDPIQEgXV) for running Colabs.

<small>**(Optional) CONSENT.** <mark>[ X ]</mark> We consent to sharing our Colab (after the assignment ends) with other students/instructors for educational purposes. We understand that sharing is optional and this decision will not affect our grade in any way. <font color=gray><i>(If ok with sharing your Colab for educational purposes, leave "X" in the check box.)</i></font></small>

In [ ]:
%%time
%reset -f
from IPython.core.interactiveshell import InteractiveShell as IS; IS.ast_node_interactivity = "all"
import pandas as pd, time, numpy as np
from sklearn.model_selection import train_test_split
from sklearn.datasets import fetch_openml
from sklearn.multioutput import MultiOutputRegressor
from sklearn.linear_model import Ridge
ToCSV = lambda df, fname: df.round(2).to_csv(f'{fname}.csv', index_label='id') # rounds values to 2 decimals

class Timer():
  def __init__(self, lim:'RunTimeLimit'=60): self.t0, self.lim, _ = time.time(), lim, print('timer started')
  def ShowTime(self):
    msg = f'Runtime is {time.time()-self.t0:.0f} sec'
    print(f'\033[91m\033[1m' + msg + f' > {self.lim} sec limit!!!\033[0m' if (time.time()-self.t0-1) > self.lim else msg)

np.set_printoptions(linewidth=10000, precision=4, edgeitems=20, suppress=True)
pd.set_option('display.max_rows', 100, 'display.max_columns', 100, 'display.max_colwidth', 100, 'display.precision', 2, 'display.max_rows', 4)

db = fetch_openml('BNG(auto_price)')   # load databunch (dictionary)
tX = pd.DataFrame(db['data'], columns=db['feature_names'])
tX.symboling = tX.symboling.astype('float')
tX['price'] = db['target']
YCols = ['city-mpg','highway-mpg','price']  # 3 targets
tY = tX[YCols]
tX.drop(YCols, axis=1, inplace=True)
# tY = pd.Series(db['target'], name='price')
tX, vX, tY, DO_NOT_USE = train_test_split(tX, tY, train_size=0.7, random_state=0, shuffle=True)
# ToCSV(DO_NOT_USE, 'testY')   # Students cannot use these test values
del DO_NOT_USE
tX
tY
tmr = Timer() # runtime limit (in seconds). Add all of your code after the timer

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


timer started
CPU times: user 23.8 s, sys: 1.74 s, total: 25.5 s
Wall time: 58.5 s


In [ ]:
tmr = Timer()

timer started


<hr color=red>

<font size=5>⏳</font> <strong><font color=orange size=5>Your Code, Documentation, Ideas and Timer - All Start Here...</font></strong>

**Student's Section** (between ⏳ symbols): add your code and documentation here.

## **Task 1. Preprocessing Pipeline**

Explain elements of your preprocessing pipeline i.e. feature engineering, subsampling, clustering, dimensionality reduction, etc.
1. Why did you choose these elements? (Something in EDA, prior experience,...? Btw, EDA is not required)
1. How do you evaluate the effectiveness of these elements?
1. What else have you tried that worked or didn't?

**Student's answer:**
**Student's answer:**
1. There were some of vital elements chosen such as MinMaxScaller (optimizing data in the given range and trying to get them equally distributed to get a higher result on the model perfomance), Polynomial features (in the actual version of the collab the Polynomial features with degree 3 are used. This method adds the most important feature(the one that was chosen throught the feature importance algorithm) to the dataset to make the prediction more precise)
2. All the elements that were used in combination with the chosen model, are pretty useful making the model work faster with more precise output.
3. The method of log application was tried to make the features distributed more equally. So, as the algorithm to find the best of features to apply log. Also, together with other methods the feature importance was used to find the most valuable features to the prediction and further work with those (devision of the most important feature by the less important one to get a new feature more important than the previous, features to be divided may be correlated, so the correlation matrix was used). All in all, the other versions of PCA and scaller were used with lower model perfomance than it is now.

In [ ]:
#Example of code for the correlation matrix

#corr = merged_data.corr()
#plt.figure(figsize=(12, 10))

#sns.heatmap(corr[(corr >= 0.4) | (corr <= -0.4)],
#            cmap='coolwarm', vmax=1.0, vmin=-1.0, linewidths=0.1,
#            annot=True, annot_kws={"size": 8}, square=True);

In [ ]:
#Example of code for the feature importance

#from sklearn.ensemble import RandomForestRegressor
#import numpy as np
#import matplotlib.pyplot as plt

#train_set_feat = train_set[:60000]
#train_set_pred_feat = train_set_pred[:60000]

#md = RandomForestRegressor()
#md.fit(train_set_feat, train_set_pred_feat)

#feature_importances = md.feature_importances_

#feature_names = [
#    'symboling', 'normalized-losses', 'wheel-base', 'length', 'width',
#    'height', 'curb-weight', 'engine-size', 'bore', 'stroke',
#    'compression-ratio', 'horsepower', 'peak-rpm'
#]

#sorted_indices = np.argsort(feature_importances)[::-1]
#sorted_feature_importances = feature_importances[sorted_indices]
#sorted_feature_names = [feature_names[i] for i in sorted_indices]

#for i, (name, importance) in enumerate(zip(sorted_feature_names, sorted_feature_importances), start=1):
#    print(f'{i}. {name}: {importance}')

## **Task 2. Modeling Approach**
Explain your modeling approach, i.e. ideas you tried and why you thought they would be helpful.

1. How did these decisions guide you in modeling?
1. How do you evaluate the effectiveness of these elements?
1. What else have you tried that worked or didn't?

**Student's answer:**

1. The Ridge() model was chosen for that task as it had the best performance on the initial dataset from all models we’ve tried.
2. For my data elements usage and model choice seem pretty effective as after all manipulations for the “Ridge” model gave the highest score of all models tried. Thus, with best performance and lowest time taken, the “Ridge” model was chosen.
3. The models that were also tried: KNeighborsClassifier, Linear Regression, RidgeCV, Lasso. Using them the code worked, however, it was less effective, taking more time than the final choice, so that models were rejected in favor of "Ridge" model.


<font color=green><h3><b>$1$. Data loading</b><h3>

In [ ]:
X_train = tX.copy()
final = vX.copy()

<font color=green><h3><b>$2$. Scaling data</b><h3>

In [ ]:
from sklearn.preprocessing import MinMaxScaler

mms = MinMaxScaler()
mms.fit(X_train)

X_train = pd.DataFrame(mms.transform(X_train), columns=X_train.columns)

X_final = pd.DataFrame(mms.transform(final), columns=X_train.columns)

MinMaxScaler()

<font color=green><h3><b>$3$. Adding polynomial features</b><h3>

In [ ]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=3)    # add non-linear features (powers and interactions)
X_train = poly.fit_transform(X_train.select_dtypes(include=np.number))  # create object on training set
X_final  = poly.transform(X_final.select_dtypes(include=np.number))       # apply the same object to test set

<font color=green><h3><b>$4$. Fitting model and asessing results </b><h3>

In [ ]:
m = MultiOutputRegressor(Ridge(random_state=42)).fit(X_train , tY)  # always seed RNG for reproducibility
print(f'In-sample R^2 = {m.score(X_train ,tY):.4f}')

In-sample R^2 = 0.3721


<font color=green><h3><b>Archive of results</b><h3>

0.3245 - baseline

0.3248 - test = 0.3, rs = 42

0.3253 - test = 0.3, rs = 42, norm

0.3721 - full size, mms, deg=3, ridge




<font color=green><h3><b>Predicting</b><h3>

In [ ]:
pY = pd.DataFrame(m.predict(X_final), index=vX.index, columns=YCols)
ToCSV(pY, 'Auto_baseline3')
pY

city-mpg  highway-mpg     price
157105     27.42        33.06   8325.56
374554     24.24        30.50  11242.57
...          ...          ...       ...
952836     25.16        30.52  13151.86
374097     30.11        35.27   6474.60

[300000 rows x 3 columns]

# **References:**

1. Remember to cite your sources here as well! At the least, your textbook should be cited. Google Scholar allows you to effortlessly copy/paste an APA citation format for books and publications. Also cite StackOverflow, package documentation, and other meaningful internet resources to help your peers learn from these (and to avoid plagiarism claims).

<font color=black><b>The following resources were used:</b></font>

1. https://hastie.su.domains/ISLR2/ISLRv2_corrected_June_2023.pdf
2. https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html
3. https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html
4. https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.RadiusNeighborsClassifier.html#sklearn.neighbors.RadiusNeighborsClassifier
5. https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA
6. https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#:~:text=Transform%20features%20by%20scaling%20each,e.g.%20between%20zero%20and%20one.
7. Seminar notes
8. Machine_Learning_with_Pytorch_and_Scikit_Learn_2022_Sebastian_Raschka.pdf (book)

<font size=5>⌛</font> <strong><font color=orange size=5>Do not exceed competition's runtime limit!</font></strong>

<hr color=red>


In [ ]:
tmr.ShowTime()    # measure Colab's runtime. Do not remove. Keep as the last cell in your notebook.

Runtime is 55 sec


## 💡**Starter Ideas**

1. Tune model hyperparameters and try different allowed models
1. Try to linear and non-linear feature normalization: shift/scale, log, divide features by features (investigate scatterplot matrix)
1. Try higher order feature interactions and polynomial features on a small subsample. Then identify key features or select key principal components. The final model can be trained on a larger or even full training sample. You can use [PCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html) to reduce the feature set
1. Do a thorough EDA: look for feature augmentations that result in linear decision boundaries between pairs of classes.
1. Evaluate predictions and focus on poorly predicted "groups":
  1. Strongest errors. E.g. the model is very confident about the wrong label
1. Do scatter plots show piecewise linear shape? Can a separate linear model be used on each support, or can the pattern be linearized via transformations?
1. Try modeling each output separately from inputs or from a other modeled output
1. Try stepwise selection and regularization and remove "unimportant" features from final model